external-crates/move/crates/move-compiler/src/typing/translate.rs

```diff
@@ -155,28 +148,50 @@ fn extract_macros(
                 None
             }
         })
-    });
-
-    context.set_macros(all_macro_definitions);
+    })
 }
 
 fn modules(
-    context: &mut Context,
+    compilation_env: &CompilationEnv,
+    info: &mut NamingProgramInfo,
+    all_macro_definitions: &UniqueMap<ModuleIdent, UniqueMap<FunctionName, N::Sequence>>,
     mut modules: UniqueMap<ModuleIdent, N::ModuleDefinition>,
 ) -> UniqueMap<ModuleIdent, T::ModuleDefinition> {
-    let mut all_new_friends = BTreeMap::new();
+    let global_use_funs = global_use_funs(info);
     // We validate the syntax methods first so that processing syntax method forms later are
     // better-typed. It would be preferable to do this in naming, but the typing machinery makes it
     // much easier to enforce the typeclass-like constraints. We also update the program info to
     // reflect any changes that happened.
-    for (key, mdef) in modules.key_cloned_iter_mut() {
-        validate_syntax_methods(context, &key, mdef);
-        context
-            .modules
-            .set_module_syntax_methods(key, mdef.syntax_methods.clone());
+    for (mident, mdef) in modules.key_cloned_iter_mut() {
+        let context = ModuleContext::new(
+            compilation_env,
+            info,
+            &global_use_funs,
+            all_macro_definitions,
+        );
+        validate_syntax_methods(&mut context.new_module_member(), &mident, mdef);
+    }
+    for (mident, mdef) in modules.key_cloned_iter() {
+        info.set_module_syntax_methods(mident, mdef.syntax_methods.clone());
     }
-    let mut typed_modules = modules.map(|ident, mdef| {
-        let (typed_mdef, new_friends) = module(context, ident, mdef);
+    let typed_modules = Mutex::new(UniqueMap::new());
+    let all_new_friends = Mutex::new(BTreeMap::new());
+    let used_module_members = Mutex::new(BTreeMap::new());
```

Comment:
I tried using `.fold` and `.reduce` instead of this lock based approach, but the lock based approach won out. 

I imagine there might be something even better here where we push to a lockless data structure, and then gather them all in the map at the end.